worker/secretdrainworker/worker.go

```diff
@@ -0,0 +1,237 @@
+)
+
+// logger is here to stop the desire of creating a package level logger.
+// Don't do this, instead use the one passed as manifold config.
+type logger interface{}
+
+var _ logger = struct{}{}
+
+// Logger represents the methods used by the worker to log information.
+type Logger interface {
+	Debugf(string, ...interface{})
+	Warningf(string, ...interface{})
+	Infof(string, ...interface{})
+}
+
+// SecretsDrainFacade instances provide a set of API for the worker to deal with secret drain process.
+type SecretsDrainFacade interface {
+	WatchSecretBackendChanged() (watcher.NotifyWatcher, error)
+	GetSecretsToDrain() ([]coresecrets.SecretMetadataForDrain, error)
+	ChangeSecretBackend([]secretsdrain.ChangeSecretBackendArg) (secretsdrain.ChangeSecretBackendResult, error)
+}
+
+// Config defines the operation of the Worker.
+type Config struct {
+	SecretsDrainFacade
+	Logger Logger
+
+	SecretsBackendGetter func() (jujusecrets.BackendsClient, error)
+}
+
+// Validate returns an error if config cannot drive the Worker.
+func (config Config) Validate() error {
+	if config.SecretsDrainFacade == nil {
+		return errors.NotValidf("nil SecretsDrainFacade")
+	}
+	if config.Logger == nil {
+		return errors.NotValidf("nil Logger")
+	}
+	if config.SecretsBackendGetter == nil {
+		return errors.NotValidf("nil SecretsBackendGetter")
+	}
+	return nil
+}
+
+// NewWorker returns a secretdrainworker Worker backed by config, or an error.
+func NewWorker(config Config) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &Worker{config: config}
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	})
+	return w, errors.Trace(err)
+}
+
+// Worker drains secrets to the new backend when the model's secret backend has changed.
+type Worker struct {
+	catacomb catacomb.Catacomb
+	config   Config
+}
+
+// TODO(secrets): user created secrets should be drained on the controller because they do not have an owner unit!
+
+// Kill is defined on worker.Worker.
+func (w *Worker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *Worker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *Worker) loop() (err error) {
+	watcher, err := w.config.SecretsDrainFacade.WatchSecretBackendChanged()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := w.catacomb.Add(watcher); err != nil {
+		return errors.Trace(err)
+	}
+
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return errors.Trace(w.catacomb.ErrDying())
+		case _, ok := <-watcher.Changes():
+			if !ok {
+				return errors.New("secret backend changed watch closed")
+			}
+
+			secrets, err := w.config.SecretsDrainFacade.GetSecretsToDrain()
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if len(secrets) == 0 {
+				continue
+			}
+			backends, err := w.config.SecretsBackendGetter()
+			if err != nil {
+				return errors.Trace(err)
+			}
```

Comment:
Can we do this up above outside the for loop?

Reply:
We may have added new backends after the worker started.